sklearn/utils/estimator_checks.py

```diff
@@ -1191,6 +1192,56 @@ def check_estimators_pickle(name, estimator_orig):
         assert_allclose_dense_sparse(result[method], unpickled_result)
 
 
+def check_pairwise_estimator_tag(name, estimator_orig):
+    attributes_to_check = [('metric', 'precomputed'),
+                           ('affinity', 'precomputed'),
+                           ('kernel', 'precomputed')]
+
+    # Using iris as sample data
+    iris = load_iris()
+    X, y_ = iris.data, iris.target
+    distance_matrix = pairwise_distances(X)
+
+    for attribute, attribute_value in attributes_to_check:
+        # Check to see if attribute value is supported by estimator
+        if getattr(estimator_orig, attribute, None) is not None:
+            continue
+        try:
+            # Construct new object of estimator with desired attribute value
+            modified_estimator = estimator_orig.__class__(
+                                    **{
+                                        attribute: attribute_value
+                                      })
+            # Not all estimators validate parameters, so check fit()
+            modified_estimator.fit(X=distance_matrix, y=y_)
+        except (TypeError, ValueError, KeyError):
+            # Estimator does not support given attribute value
+            continue
+
+        # Also check to see if non-square distance matrix raises an error
+        try:
+            non_square_distance = distance_matrix[:, :-1]
+            if getattr(modified_estimator, 'fit_predict', None) is not None:
+                modified_estimator.fit_predict(X=non_square_distance, y=y_)
+            elif (getattr(modified_estimator, 'fit_transform', None)
+                    is not None):
+                modified_estimator.fit_transform(X=non_square_distance, y=y_)
+            elif getattr(modified_estimator, 'fit', None) is not None:
+                modified_estimator.fit(X=non_square_distance, y=y_)
+                if getattr(modified_estimator, 'predict', None) is not None:
+                    modified_estimator.predict(X=X)
+        except ValueError:
+            # Check if estimator defines _pairwise attribute
+            assert_not_equal(getattr(modified_estimator, '_pairwise', None),
```

Comment:
Can we not check as far as "_pairwise should be True iff 'precomputed'"?

Reply:
It's possible to check both ways, but I'm not sure if this test is a necesary condition for `_pairwise` (though it is sufficient). I implemented the necessary check (you can find it here: https://github.com/gokart23/scikit-learn/commit/04209a419718bf2139c93684ee976ac63a72680b) but there turned out to be a couple of failing estimators:
1. `sklearn.preprocessing.data.KernelCenterer` : `_pairwise` attribute added in commit https://github.com/scikit-learn/scikit-learn/commit/2242b1b2ea715161f87307060104a8a0141e5077
2. `sklearn.neighbors.unsupervised.NearestNeighbors`: `_pairwise` attribute added in commit https://github.com/scikit-learn/scikit-learn/commit/f4fc2753c3177dafbd49692ce05a324e65e57d8f

I'm not sure if these are valid failures, or how to go about fixing these failures, which is why I haven't included them in the PR as yet. What do you think?